resources/sqlite/src/conn_opts.rs

```diff
@@ -0,0 +1,270 @@
+use std::str::FromStr;
+use std::time::Duration;
+use std::{borrow::Cow, path::PathBuf};
+
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+use sqlx::sqlite::SqliteConnectOptions;
+
+mod auto_vacuum;
+pub use auto_vacuum::*;
+
+mod journal_mode;
+pub use journal_mode::*;
+
+mod locking_mode;
+pub use locking_mode::*;
+
+mod synchronous;
+pub use synchronous::*;
+use tracing::debug;
+
+/// Options to configure the SQLite database mirroring sqlx's [`SqliteConnectOptions`](https://docs.rs/sqlx/latest/sqlx/sqlite/struct.SqliteConnectOptions.html)
+/// for the options it exposes, see their docs for reference.
+///
+/// Construction of the full connection string is handled internally for security reasons and defaults to creating a
+/// file-based database named `default_db.sqlite` with `create_if_missing == true`. Use the `filename` and/or
+/// `in_memory` methods to configure the type of database created.
+///
+/// Note that Shuttle does currently not support the `collation`, `thread_name`, `log_settings`, `pragma`, `extension`,
+/// `shared_cache` options.
+#[derive(Debug, Deserialize, Serialize)]
+pub struct SQLiteConnOpts {
+    // Used for constructing the full connection string internally in `try_from`.
+    pub(crate) storage_path: PathBuf,
+    // Mirrored options from the original.
+    pub(crate) filename: Cow<'static, Path>,
+    pub(crate) in_memory: bool,
+    pub(crate) read_only: bool,
+    pub(crate) create_if_missing: bool,
+    pub(crate) statement_cache_capacity: usize,
+    pub(crate) busy_timeout: Duration,
+    pub(crate) immutable: bool,
+    pub(crate) vfs: Option<Cow<'static, str>>,
+
+    pub(crate) pragmas: IndexMap<Cow<'static, str>, Option<Cow<'static, str>>>,
+
+    pub(crate) command_channel_size: usize,
+    pub(crate) row_channel_size: usize,
+
+    pub(crate) serialized: bool,
+}
+
+impl Default for SQLiteConnOpts {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SQLiteConnOpts {
+    pub fn new() -> Self {
+        let mut pragmas: IndexMap<Cow<'static, str>, Option<Cow<'static, str>>> = IndexMap::new();
+
+        pragmas.insert("key".into(), None);
+
+        pragmas.insert("cipher_plaintext_header_size".into(), None);
+
+        pragmas.insert("cipher_salt".into(), None);
+
+        pragmas.insert("kdf_iter".into(), None);
+
+        pragmas.insert("cipher_kdf_algorithm".into(), None);
+
+        pragmas.insert("cipher_use_hmac".into(), None);
+
+        pragmas.insert("cipher_compatibility".into(), None);
+
+        pragmas.insert("cipher_page_size".into(), None);
+
+        pragmas.insert("cipher_hmac_algorithm".into(), None);
+
+        pragmas.insert("page_size".into(), None);
+
+        pragmas.insert("locking_mode".into(), None);
+
+        pragmas.insert("journal_mode".into(), None);
+
+        pragmas.insert("foreign_keys".into(), Some("ON".into()));
+
+        pragmas.insert("synchronous".into(), None);
+
+        pragmas.insert("auto_vacuum".into(), None);
```

Comment:
Do we need to insert all of these here, or can we just do it if/when they are set?

Reply:
Hm, so following the principle of mirroring the original config, I put it here to make sure that we're not missing any interactions that we don't know about, e.g. making sure that all the original defaults are set. Take the `foreign_keys` pragma for example:

When the `foreign_keys` method is being used, it sets `ON` or `OFF` but the `ON` variant is being set as the default, what happens if it's not set I don't really know. True, it is a bit verbose but this way we have a catch-all way of making sure that the config is initialised correctly without knowing of all the internal interdependencies that there may be.